File: src/OracleScry.Domain/Entities/Card.cs
using OracleScry.Domain.ValueObjects;

namespace OracleScry.Domain.Entities;

/// <summary>
/// Card entity representing an MTG card with full Scryfall schema.
/// This is the aggregate root for card-related data.
/// </summary>
public class Card : BaseEntity
{
    // Scryfall Core Fields
    public Guid ScryfallId { get; set; }
    public string Object { get; set; } = "card";
    public string Uri { get; set; } = string.Empty;
    public string ScryfallUri { get; set; } = string.Empty;
    public string PrintsSearchUri { get; set; } = string.Empty;
    public string RulingsUri { get; set; } = string.Empty;

    // Identification Fields
    public int? ArenaId { get; set; }
    public int? MtgoId { get; set; }
    public int? MtgoFoilId { get; set; }
    public List<int> MultiverseIds { get; set; } = [];
    public int? TcgplayerId { get; set; }
    public int? TcgplayerEtchedId { get; set; }
    public int? CardmarketId { get; set; }
    public Guid? OracleId { get; set; }

    // Gameplay Fields
    public string Name { get; set; } = string.Empty;
    public string Lang { get; set; } = "en";
    public string Layout { get; set; } = string.Empty;
    public string? ManaCost { get; set; }
    public decimal Cmc { get; set; }
    public string TypeLine { get; set; } = string.Empty;
    public string? OracleText { get; set; }
    public string? Power { get; set; }
    public string? Toughness { get; set; }
    public string? Loyalty { get; set; }
    public string? Defense { get; set; }
    public List<string> Colors { get; set; } = [];
    public List<string> ColorIdentity { get; set; } = [];
    public List<string>? ColorIndicator { get; set; }
    public List<string>? ProducedMana { get; set; }
    public List<string> Keywords { get; set; } = [];
    public bool Reserved { get; set; }
    public int? EdhrecRank { get; set; }
    public int? PennyRank { get; set; }
    public bool? GameChanger { get; set; }
    public string? HandModifier { get; set; }
    public string? LifeModifier { get; set; }

    // Print Fields
    public string? Artist { get; set; }
    public List<Guid> ArtistIds { get; set; } = [];
    public bool Booster { get; set; }
    public string BorderColor { get; set; } = string.Empty;
    public Guid CardBackId { get; set; }
    public string CollectorNumber { get; set; } = string.Empty;
    public bool? ContentWarning { get; set; }
    public bool Digital { get; set; }
    public List<string> Finishes { get; set; } = [];
    public string? FlavorName { get; set; }
    public string? FlavorText { get; set; }
    public List<string>? FrameEffects { get; set; }
    public string Frame { get; set; } = string.Empty;
    public bool FullArt { get; set; }
    public List<string> Games { get; set; } = [];
    public bool HighresImage { get; set; }
    public Guid? IllustrationId { get; set; }
    public string ImageStatus { get; set; } = string.Empty;
    public bool Oversized { get; set; }
    public string? PrintedName { get; set; }
    public string? PrintedText { get; set; }
    public string? PrintedTypeLine { get; set; }
    public bool Promo { get; set; }
    public List<string>? PromoTypes { get; set; }
    public string Rarity { get; set; } = string.Empty;
    public DateTime ReleasedAt { get; set; }
    public bool Reprint { get; set; }
    public string SetCode { get; set; } = string.Empty;
    public Guid SetId { get; set; }
    public string SetName { get; set; } = string.Empty;
    public string SetType { get; set; } = string.Empty;
    public string SetUri { get; set; } = string.Empty;
    public string SetSearchUri { get; set; } = string.Empty;
    public string ScryfallSetUri { get; set; } = string.Empty;
    public bool StorySpotlight { get; set; }
    public bool Textless { get; set; }
    public bool Variation { get; set; }
    public Guid? VariationOf { get; set; }
    public string? SecurityStamp { get; set; }
    public string? Watermark { get; set; }
    public List<int>? AttractionLights { get; set; }

    // Complex Objects (JSON Columns)
    public CardImageUris? ImageUris { get; set; }
    public CardPrices Prices { get; set; } = new();
    public CardLegalities Legalities { get; set; } = new();
    public Dictionary<string, string>? PurchaseUris { get; set; }
    public Dictionary<string, string> RelatedUris { get; set; } = new();
    public CardPreview? Preview { get; set; }

    // Navigation Properties (Separate Tables)
    public ICollection<CardFace> CardFaces { get; set; } = [];
    public ICollection<RelatedCard> AllParts { get; set; } = [];
}

File: frontend/src/types/card.ts
/**
 * Card image URIs for different sizes
 */
export interface CardImageUris {
  small?: string;
  normal?: string;
  large?: string;
  png?: string;
  artCrop?: string;
  borderCrop?: string;
}

/**
 * Card prices from different sources
 */
export interface CardPrices {
  usd?: string;
  usdFoil?: string;
  usdEtched?: string;
  eur?: string;
  eurFoil?: string;
  tix?: string;
}

/**
 * Card legality in various formats
 */
export interface CardLegalities {
  standard?: string;
  future?: string;
  historic?: string;
  timeless?: string;
  gladiator?: string;
  pioneer?: string;
  explorer?: string;
  modern?: string;
  legacy?: string;
  pauper?: string;
  vintage?: string;
  penny?: string;
  commander?: string;
  oathbreaker?: string;
  standardbrawl?: string;
  brawl?: string;
  alchemy?: string;
  paupercommander?: string;
  duel?: string;
  oldschool?: string;
  premodern?: string;
  predh?: string;
}

/**
 * Card face for multi-faced cards
 */
export interface CardFaceDto {
  id: string;
  name: string;
  manaCost?: string;
  typeLine?: string;
  oracleText?: string;
  flavorText?: string;
  colors: string[];
  colorIndicator: string[];
  power?: string;
  toughness?: string;
  loyalty?: string;
  defense?: string;
  artist?: string;
  artistId?: string;
  illustrationId?: string;
  imageUris?: CardImageUris;
  faceIndex: number;
}

/**
 * Related card reference
 */
export interface RelatedCardDto {
  id: string;
  scryfallId: string;
  component: string;
  name: string;
  typeLine?: string;
  uri?: string;
}

/**
 * Card preview information
 */
export interface CardPreview {
  previewedAt?: string;
  sourceUri?: string;
  source?: string;
}

/**
 * Full card details DTO
 */
export interface CardDto {
  id: string;
  scryfallId: string;
  oracleId?: string;
  mtgoId?: number;
  mtgoFoilId?: number;
  tcgplayerId?: number;
  tcgplayerEtchedId?: number;
  cardmarketId?: number;
  arenaId?: number;
  name: string;
  lang: string;
  releasedAt?: string;
  uri?: string;
  scryfallUri?: string;
  layout: string;
  highresImage: boolean;
  imageStatus?: string;
  imageUris?: CardImageUris;
  manaCost?: string;
  cmc: number;
  typeLine?: string;
  oracleText?: string;
  power?: string;
  toughness?: string;
  loyalty?: string;
  defense?: string;
  colors: string[];
  colorIdentity: string[];
  colorIndicator: string[];
  keywords: string[];
  allParts: RelatedCardDto[];
  cardFaces: CardFaceDto[];
  legalities?: CardLegalities;
  games: string[];
  reserved: boolean;
  foil: boolean;
  nonfoil: boolean;
  finishes: string[];
  oversized: boolean;
  promo: boolean;
  reprint: boolean;
  variation: boolean;
  setId?: string;
  set: string;
  setName?: string;
  setType?: string;
  setUri?: string;
  setSearchUri?: string;
  scryfallSetUri?: string;
  rulingsUri?: string;
  printsSearchUri?: string;
  collectorNumber?: string;
  digital: boolean;
  rarity?: string;
  flavorText?: string;
  cardBackId?: string;
  artist?: string;
  artistIds: string[];
  illustrationId?: string;
  borderColor?: string;
  frame?: string;
  frameEffects: string[];
  securityStamp?: string;
  fullArt: boolean;
  textless: boolean;
  booster: boolean;
  storySpotlight: boolean;
  promoTypes: string[];
  edhrecRank?: number;
  pennyRank?: number;
  prices?: CardPrices;
  relatedUris?: Record<string, string>;
  purchaseUris?: Record<string, string>;
  preview?: CardPreview;
  producedMana: string[];
  lifeModifier?: string;
  handModifier?: string;
  attractionLights: number[];
  contentWarning: boolean;
  flavorName?: string;
  watermark?: string;
  importedOn: string;
  lastUpdatedOn: string;
}

/**
 * Card summary for list views
 */
export interface CardSummaryDto {
  id: string;
  scryfallId: string;
  name: string;
  manaCost?: string;
  cmc: number;
  typeLine?: string;
  rarity?: string;
  set: string;
  setName?: string;
  collectorNumber?: string;
  imageUri?: string;
  prices?: CardPrices;
}

/**
 * Card filter parameters for search
 */
export interface CardFilterDto {
  search?: string;
  name?: string;
  set?: string;
  rarity?: string;
  colors?: string[];
  colorIdentity?: string[];
  types?: string[];
  minCmc?: number;
  maxCmc?: number;
  format?: string;
  purposes?: string[];
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDescending?: boolean;
}

/**
 * Format legality values
 */
export type LegalityStatus = 'legal' | 'not_legal' | 'restricted' | 'banned';

/**
 * Card rarity values
 */
export type CardRarity = 'common' | 'uncommon' | 'rare' | 'mythic' | 'special' | 'bonus';

/**
 * MTG color codes
 */
export type MtgColor = 'W' | 'U' | 'B' | 'R' | 'G';
